web/static/components/provenance/utils.js

```diff
@@ -0,0 +1,145 @@
+import * as util from "../../util.js";
+import { DEFAULTS } from "./state.js";
+
+function createNode(item) {
+    const node = {
+        id: item.id,
+        doi: item.doi,
+        size: item.size,
+        notes: item.notes,
+        inhErr: item.inhErr,
+        locked: item.locked,
+        links: [],
+        nodeSize: DEFAULTS.NODE_SIZE,
+        labelSize: DEFAULTS.LABEL_SIZE,
+    };
+
+    makeLabel(node, item);
+
+    if (item.gen !== undefined) {
+        node.row = item.gen;
+        node.col = 0;
+    }
+
+    return node;
+}
+
+function makeLabel(node, item) {
+    if (item.alias) {
+        node.label = item.alias;
+    } else node.label = item.id;
+
+    node.label += util.generateNoteSpan(item, true);
+}
+
+function graphCountConnected(a_node, a_visited, a_from) {
+    let count = 0;
+
+    if (a_visited.indexOf(a_node.id) < 0 && !a_node.prune) {
+        a_visited.push(a_node.id);
+        count++;
+        let link, dest;
+        for (let i in a_node.links) {
+            link = a_node.links[i];
+            if (link !== a_from) {
+                dest = link.source === a_node ? link.target : link.source;
+                count += graphCountConnected(dest, a_visited, link);
+            }
+        }
+    }
+
+    return count;
+}
+
+// Depth-first-search to required nodes, mark for pruning
+function graphPruneCalc(a_node, a_visited, a_source) {
```

Comment:
Is this being used as part of the hide command or the collapse one? Would probably be good to put it in the comments. I know we discussed these features and how they are big confusing as it is.